multicluster/controllers/multicluster/label_identity_controller.go

```diff
@@ -0,0 +1,347 @@
+		client.Client
+		Scheme           *runtime.Scheme
+		commonAreaGetter RemoteCommonAreaGetter
+		remoteCommonArea commonarea.RemoteCommonArea
+		// labelMutex prevents concurrent access to labelToPodsCache and podLabelCache.
+		// It also prevents concurrent updates to ResourceExports.
+		labelMutex sync.Mutex
+		// labelToPodsCache stores mapping from label identities to Pods that have this label identity.
+		labelToPodsCache    map[string]sets.String
+		resourceExportMutex sync.Mutex
+		// labelAddsInProgress keeps track of label identities whose corresponding ResourceExports are
+		// currently being created in the CommonArea. It is protected by resourceExportMutex.
+		labelAddsInProgress sets.String
+		// labelDeletesInProgress keeps track of label identities whose corresponding ResourceExports are
+		// currently being deleted. It is protected by resourceExportMutex.
+		labelDeletesInProgress sets.String
+		// podLabelCache stores mapping from Pods to their label identities.
+		podLabelCache  map[string]string
+		localClusterID string
+	}
+)
+
+func NewLabelIdentityReconciler(
+	client client.Client,
+	scheme *runtime.Scheme,
+	commonAreaGetter RemoteCommonAreaGetter) *LabelIdentityReconciler {
+	return &LabelIdentityReconciler{
+		Client:                 client,
+		Scheme:                 scheme,
+		commonAreaGetter:       commonAreaGetter,
+		labelToPodsCache:       map[string]sets.String{},
+		labelAddsInProgress:    sets.NewString(),
+		labelDeletesInProgress: sets.NewString(),
+		podLabelCache:          map[string]string{},
+	}
+}
+
+// +kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=resourceexports,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch
+// +kubebuilder:rbac:groups="",resources=namespaces,verbs=get;list;watch
+func (r *LabelIdentityReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	klog.V(2).InfoS("Reconciling Pod/Namespace for label identity", "pod/ns", req.NamespacedName)
+	var commonArea commonarea.RemoteCommonArea
+	var err error
+	commonArea, r.localClusterID, err = r.commonAreaGetter.GetRemoteCommonAreaAndLocalID()
+	if commonArea == nil {
+		return ctrl.Result{Requeue: true}, err
+	}
+	r.remoteCommonArea = commonArea
+	var pod v1.Pod
+	var ns v1.Namespace
+
+	if err := r.Client.Get(ctx, req.NamespacedName, &pod); err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.V(2).InfoS("Pod is deleted", "pod", req.NamespacedName)
+			return ctrl.Result{}, r.onPodDelete(ctx, req.NamespacedName.String())
+		}
+		return ctrl.Result{}, err
+	}
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: req.Namespace}, &ns); err != nil {
+		klog.ErrorS(err, "Cannot get corresponding Namespace of the Pod", "pod", req.NamespacedName)
+		return ctrl.Result{}, err
+	}
+	normalizedLabel := getNormalizedLabel(ns.Labels, pod.Labels, ns.Name)
+	requeueIfLabelResExportUpdateInProgress := func() bool {
```

Comment:
Does this way work? Another worker may start to add/remove the label after the check. If we go this requeue approach, we need to mark the state right after the check passed (i.e. state check and update should be atomic).

Do you know the (mem) overhead of Mutex in Golang? If the overhead is not big, maybe simpler to save a mutex into the label cache) and use that to protect add/delete of each label.

Reply:
I am not sure I follow the first part. We use the `resourceExportMutex` in a way such that any label identity ResExport currently being created/deleted are guaranteed to be in the "in progress" set. Once the creation/deletion is done they are immediately removed from the set. Any worker who holds the label key in between it was added to the set and removed from the set will need to be re-queued. I don't feel that the check and update need to be atomic, as that will be blocking and not much difference from locking the whole function.

Reply:
I meant here we just check labelAdds/DeletesInProgress, but we do not change their state. It is possible the check passed at this time point, but another worker is processing label add/delete too will save the label to labelAdds/DeletesInProgress, for example before onPodCreateOrUpdate() is called by this worker.

Reply:
I think I understood your point now and it seems this case can be solved by having the `requeueIfLabelResExportUpdateInProgress` acquire both `resourceExportMutex` and `labelMutex`, in that specific order. I have checked all possible scenarios and don't think a deadlock can potentially occur.  Please let me know if you think otherwise.

I did try to implement mutex per label but found out that it was really tricky as well. For update cases, we need to lock both the new label and the original label, and there's a sequence to it as well since we can only get the original label after looking at the cache. This makes implementation look quite messy, though I think maintaining a mutex per label shouldn't be too much of a overhead

Reply:
I did not get your idea of acquiring both mutexes. Probably I will read your code later to understand. But I was thinking we should **check and update `labelAdds/DeletesInProgress` atomically** after we figure out which label to add/delete; we cannot just check and requeue events here. Maybe we just need one mutex of `labelMutex` for that.

About per label mutex, I thought after we figure out which label to add, when can then lock it and check it again to decide whether a creation is needed; and we do the same for delete. Why we need to lock both label to add and label to delete the same time?

Last, I thought over about the problem. I feel the race between add and delete that we are trying to solve is because we want to ignore ALREADY_EXISTS for creation. If we do not ignore that, maybe we need not synchronization for add/delete of a single label. But then we need to make sure after controller restarts we should restore all state including label caches correctly, before we progress any add events (so we are sure ALREDY_EXISTS is not due to previously created Export before restart). Do we have a way with kube builder to restore all state before watching new changes?

Let me know if you want to talk to discuss.

Reply:
@jianjuns please check the latest commit to see if the mutex per label implementation is what you expected